### ISSUE TITLE
Preprocessing bottleneck in count_point_appears_in_node (37h on Mip360 Bicycle, ID-correlated masks already available)

### ISSUE THREAD

**Issue by tehMazz** [zju3dv/InstaScene#10]
Hello,

thank you for the great work and for releasing the code! I am currently benchmarking your method against other approaches, making sure that all methods receive the same ID-correlated input masks.

While running on the Bicycle scene from Mip360, I encountered a severe bottleneck in the preprocessing stage, specifically in the function:
```
def count_point_appears_in_node(mask_list, node_frame_id_list, point_ids_list, 
                                mask_point_clouds, point_appear_in_node_matrixs)
```

With only a two-label starting masks (background and bike), the preprocessing step is estimated to take ~37 hours to complete on a NVIDIA A100 80GB 

Since in my setup the input masks are already ID-consistent across frames, would it be possible to either skip this function entirely or introduce a shortcut that bypasses the redundant preprocessing steps in such cases?

Thanks again for your work, and I’d be happy to provide more details or logs if needed.

**Comment by YZsZY** [comment]
Since our algorithm automatically applies a [DBScan process](https://github.com/zju3dv/InstaScene/blob/main/spatial_track/modules/post_process.py#L209) based on spatial distribution during preprocessing, it can avoid issues caused by under-segmentation (for example, when one mask corresponds to multiple bottles, the point cloud will be automatically split into multiple bottles based on spatial distribution). Therefore, our method is mainly suitable for extremely complex scenarios (cluttered scenes).

In the Bicycle scene, however, the background point cloud is extremely dense and spatially complex. This causes the algorithm to divide it into a large number of clusters, which makes processing very slow. If the input masks already divide the background into multiple parts, the process will speed up a lot just like the Kitchen Case in our Readme.

If your goal is only to separate foreground from background, in fact only some very lightweight processing is needed:

1) Disable the DBScan post-processing step during data preprocessing.

2) Simply use single-view contrastive learning — you don’t even need cross-view consistent ID masks.

3) If you still wish to make use of the cross-view ID-consistent mask, you can replace the following part of the code `construct_mask2gs_tracker` and `iterative_clustering` shown at [this line](https://github.com/zju3dv/InstaScene/blob/main/spatial_track/spatialtrack.py#L41), with the implementation below. This is the approach we describe in Figure 4 (Object Tracker) of our supplementary material:
```python
def graphclustering_objtracker(self):
    self.load_objtracker()
    self.post_process()

def load_objtracker(self):
    boundary_points, point_in_mask_matrix, mask_point_clouds, point_frame_matrix, global_frame_mask_list = build_point_in_mask_matrix(
        self.gaussian, self.viewcams)

    self.mask_point_clouds = mask_point_clouds
    self.point_frame_matrix = point_frame_matrix
    self.global_frame_mask_list = global_frame_mask_list  # some views are empty

    self.init_objtracker_node(mask_point_clouds, global_frame_mask_list)

def init_objtracker_node(self, mask_point_clouds, global_frame_mask_list):
    from clustering.graph.node import Node
    nodes = []
    node_idxs = set([frame_mask[1] for frame_mask in global_frame_mask_list])  # mask_id
    frame_idxs = set([frame_mask[0] for frame_mask in global_frame_mask_list])  # frame_id
    for node_idx in node_idxs:
        mask_list = []
        point_ids = set()
        contained_mask = torch.zeros((len(global_frame_mask_list),), dtype=torch.bool).cuda()
        visible_frame = torch.zeros((self.point_frame_matrix.shape[1],)).cuda()
        for global_mask_id, (frame_id, mask_id) in enumerate(global_frame_mask_list):
            if mask_id == node_idx:
                mask_list.append((frame_id, mask_id))
                point_ids.update(mask_point_clouds[f"{frame_id}_{mask_id}"])
                contained_mask[global_mask_id] = True
                visible_frame[frame_id] = 1
        node = Node(mask_list, visible_frame, contained_mask, point_ids, None, None)
        nodes.append(node)

    self.nodes = nodes
```
You can refer to the code in the spatial tracker section for implementation details.

By the way, I want to emphasize that the segmentation method we propose is designed for **much more challenging input scenarios**, where the scenes are typically indoor environments with numerous objects, repeated categories, dense arrangements, and severe occlusions. In such cases, it is very difficult to obtain ID-consistent cross-view masks directly from the input, and we also need to account for under-segmentation issues in the input masks. Be careful to distinguish this from the input usage you mentioned.
